Feature/descw 654 invoice resource deliverable

mhaswell-bcgov opened the pull request.

mhaswell-bcgov committed:
Split backend invoice_details files into invoice_deliverables and invoice_resources files.
Add controllers, models, routes, validators for invoice_deliverables and invoice_resources.
Updated picker_options model/seeds to allow contractId-specific picker option results.
Added unit tests for invoice_deliverables and invoice_resources routes.
Added support for contractId-specific picker options to frontend alongside existing projectId.
Fixed bug in TableTotalFooter causing console errors.
Added fields necessary for read/editing of invoice resources and invoice deliverables.

mhaswell-bcgov committed:
Split backend invoice_details files into invoice_deliverables and invoice_resources files.
Add controllers, models, routes, validators for invoice_deliverables and invoice_resources.
Updated picker_options model/seeds to allow contractId-specific picker option results.
Added unit tests for invoice_deliverables and invoice_resources routes.
Added support for contractId-specific picker options to frontend alongside existing projectId.
Fixed bug in TableTotalFooter causing console errors.
Added fields necessary for read/editing of invoice resources and invoice deliverables.

mhaswell-bcgov committed:
Split backend invoice_details files into invoice_deliverables and invoice_resources files.
Add controllers, models, routes, validators for invoice_deliverables and invoice_resources.
Updated picker_options model/seeds to allow contractId-specific picker option results.
Added unit tests for invoice_deliverables and invoice_resources routes.
Added support for contractId-specific picker options to frontend alongside existing projectId.
Fixed bug in TableTotalFooter causing console errors.
Added fields necessary for read/editing of invoice resources and invoice deliverables.

mhaswell-bcgov committed:
Fixed wrong type in TableTotalFooter.

mhaswell-bcgov committed:
Fixed wrong type in TableTotalFooter.

mhaswell-bcgov committed:
Fixed wrong type in TableTotalFooter.

mhaswell-bcgov committed:
Merge branch 'feature/DESCW-654-invoice-resource-deliverable' of github.com:bcgov/gdx-agreements-tracker into feature/DESCW-654-invoice-resource-deliverable

mhaswell-bcgov committed:
Fixed issues integrating with new picker system.

mhaswell-bcgov committed:
Fixed issues integrating with new picker system.

mhaswell-bcgov committed:
Merge branch 'feature/DESCW-654-invoice-resource-deliverable' of github.com:bcgov/gdx-agreements-tracker into feature/DESCW-654-invoice-resource-deliverable

mhaswell-bcgov committed:
Fixed bug with empty result sets in databaseConnection.js.
Removed unused picker_options seeds.
Fixed bugs with contract-specific queries in picker_options model.
Fixed caching bug in BudgetDisplay component.
Fixed bug in GDXMultiselect component caused by changes to picker options.
Fixed various bugs caused by picker options changes in Contracts section.

mhaswell-bcgov committed:
Updated changelog.

ShawnTurple commented on backend/src/controllers/invoice_resources.js:
```diff
@@ -0,0 +1,44 @@
+const useController = require("./useController/index.js");
+const model = require("../models/invoice_resources");
+const what = { single: "invoice resource", plural: "invoice resources" };
```
just make sure that the what having spaces don't have any side effect.

mhaswell-bcgov replied:
All the routes work so it doesn't seem to. It's just used for the error messages in this case ("invoice resource does not exist").

ShawnTurple commented on backend/src/controllers/invoice_resources.js:
```diff
@@ -0,0 +1,44 @@
+  controller.userRequires(request, what, `contracts_add_one`);
+  const invoiceId = Number(request.params.id);
+  try {
+    const result = await model.addOne(request.body, invoiceId);
```
I feel that the model shouldn't point to the same model that is used as a default in the controller addOne (useController)

mhaswell-bcgov replied:
Changed model functions from addOne() to addOneWithInvoiceId().

ShawnTurple commented on backend/src/database/databaseConnection.js:
```diff
@@ -17,9 +17,9 @@ const databaseConnection = () => {
           definition: knex.raw(`
         (
-          SELECT json_agg(${option.id}) 
+          SELECT COALESCE(json_agg(${option.id}), '[]')
```
just curious if you had an instance where this had no results (maybe with a specific contract id, and maybe with project id)

mhaswell-bcgov replied:
/projects/1154/billing is an example. If I remove the COALESCE here it fails validation (the picker options definition is supposed to be an array but json_agg returns a null). If I remove validation on the route it loads on the frontend but then when I click the select input the app crashes.

ShawnTurple commented on backend/src/models/invoice_deliverables.js:
```diff
@@ -0,0 +1,67 @@
+
+const findById = (id) => {
+  return knex
+    .select(
```
I feel that .columns() might be a better way of doing this with knex.  Maybe something we can discuss as a team.

ShawnTurple commented on backend/src/models/invoice_resources.js:
```diff
@@ -0,0 +1,96 @@
+// Get all resources by invoice id.
+const findAllByInvoiceId = (invoiceId) => {
+  return knex
+    .select(
```
again this seems like a consistency thing, i didn't know you could do this with the select statement as. well using the object, thought it was only via .columns()

ShawnTurple commented on backend/src/models/invoice_resources.js:
```diff
@@ -0,0 +1,96 @@
+const findById = (id) => {
+  return knex
+    .select(
+      "i.id",
```
we should be careful of abbreviation of tables to single letters unless it is the only table, as this one at min should be id (invoice_detail).  Just makes reading more difficult and easier to make mistakes, for what we are saving here, sometimes it might even be better to have the fully qualified table name.

ShawnTurple commented on backend/src/models/picker_options.js:
```diff
@@ -208,6 +208,26 @@ const tableLookupValues = (projectId, contractId) => {
+      description: "",
+      table: "data.contract_resource",
+      value: "contract_resource.id",
+      label: `(r.resource_last_name || ', ' || r.resource_first_name)`,
```
does this work, as how does it know '**r**' is the contract_resource table, this is not declared anywheres. -- ahh i see its in the join (queryAdditions)

ShawnTurple commented on backend/test/routes/picker_options.test.js:
```diff
@@ -12,4 +12,16 @@ testRoutes([
     capabilities: capability,
     type: routeTypes.General,
   },
+  {
```
thank you

ShawnTurple commented on frontend/src/pages/Contracts/Contract/ContractDetails/fields.tsx:
```diff
@@ -119,6 +119,7 @@ export const editFields: IEditFields[] = [
     fieldName: "supplier_id",
     fieldType: "select",
+    pickerName: "supplier_option",
     tableName: "resource",
```
you should probably remove tableName, as this is no longer required.

ShawnTurple commented on frontend/src/pages/Contracts/Contract/ContractDetails/fields.tsx:
```diff
@@ -132,6 +133,7 @@ export const editFields: IEditFields[] = [
     fieldName: "subcontractor_id",
     fieldType: "multiselect",
+    pickerName: "subcontractor_option",
     tableName: "resource",
```
remove tableName as not required.

mhaswell-bcgov replied:
Removed all the tableNames you mentioned.

ShawnTurple commented on frontend/src/pages/Contracts/Contract/InvoiceProcessing/InvoiceDeliverables/fields.tsx:
```diff
@@ -10,29 +10,50 @@ import { IEditFields } from "types";
+      fieldLabel: "Deliverables / Expense",
+      fieldName: "contract_deliverable_id",
+      fieldType: "select",
+      tableName: "contracts",
```
remove tableName

ShawnTurple commented on frontend/src/pages/Contracts/Contract/InvoiceProcessing/InvoiceResources/fields.tsx:
```diff
@@ -10,37 +10,52 @@ import { IEditFields } from "types";
-      fieldName: "contract_resource",
+      fieldName: "contract_resource_id",
       fieldType: "select",
       tableName: "contracts",
```
remove tableName

ShawnTurple commented on frontend/src/pages/Contracts/Contract/InvoiceProcessing/fields.tsx:
```diff
@@ -48,6 +48,7 @@ export const editFields: IEditFields[] = [
     fieldLabel: "Fiscal Year",
     width: "half",
+    pickerName: "fiscal_year_option",
     tableName: "generic",
```
remove tableName

ShawnTurple commented on frontend/src/pages/Contracts/Contract/Resources/fields.tsx:
```diff
@@ -32,6 +32,7 @@ export const editFields: IEditFields[] = [
     fieldName: "fiscal",
     fieldType: "select",
+    pickerName: "fiscal_year_option",
     tableName: "generic",
```
remove tableName

ShawnTurple reviewed (changes requested):
remove tableName where using pickerName, not required

mhaswell-bcgov committed:
PR feedback:
  - Renamed invoice_deliverables/resources models' addOne() to addOneWithInvoiceId().
  - Removed obsolete tableName property from fields where pickerName is used.
